### PR TITLE
Build VSCode using TS 2.4

### LINKED ISSUE #28643: Adopt TS 2.4 For VS Code Build
Adopt TS 2.4 for building VS Code

### PR DESCRIPTION
**Pull request by mjbvz**
Build VSCode with TS 2.4

Fixes #28643

### CODE REVIEW

File: build/tfs/common/publish.ts [hunk 1]
```diff
@@ -131,9 +131,11 @@ async function doesAssetExist(blobService: azure.BlobService, quality: string, b
 }
 
 async function uploadBlob(blobService: azure.BlobService, quality: string, blobName: string, file: string): Promise<void> {
-	const blobOptions = {
-		contentType: mime.lookup(file),
-		cacheControl: 'max-age=31536000, public'
+	const blobOptions: azure.BlobService.CreateBlockBlobRequestOptions = {
+		contentSettings: {
```
Comment (joaomoreno) [inline build/tfs/common/publish.ts]: Isn't this going to break something?

Comment (mjbvz) [inline build/tfs/common/publish.ts]: I believe this is the correct API for current version of the azure bindings: https://github.com/Azure/azure-storage-node/blob/master/typings/azure-storage/azure-storage.d.ts#L2778

Not sure if the properties were being applied correctly before or just being ignored

**Comment by jrieken** [comment]
There is still many compile errors (missing adoptions)

```
[11:14:35] Starting 'watch-client'...
[11:15:35] [monaco.d.ts] Starting monaco.d.ts generation
[11:15:39] [monaco.d.ts] Finished monaco.d.ts generation
[11:15:39] Error: /Users/jrieken/Code/vscode/src/vs/workbench/services/textfile/common/textFileService.ts(44,23): Class 'TextFileService' incorrectly implements interface 'ITextFileService'.
  Types of property 'revert' are incompatible.
    Type '(resource: URI, options?: IRevertOptions) => TPromise<boolean>' is not assignable to type '(resource: URI, force?: boolean) => TPromise<boolean>'.
      Types of parameters 'options' and 'force' are incompatible.
        Type 'boolean' is not assignable to type 'IRevertOptions'.
[11:15:39] Error: /Users/jrieken/Code/vscode/src/vs/workbench/services/textmodelResolver/test/textModelResolverService.test.ts(136,47): Type 'typeof TPromise' is not a valid async function return type in ES5/ES3 because it does not refer to a Promise-compatible constructor value.
[11:15:39] Error: /Users/jrieken/Code/vscode/src/vs/workbench/services/textmodelResolver/test/textModelResolverService.test.ts(136,47): Type 'typeof TPromise' is not a valid async function return type in ES5/ES3 because it does not refer to a Promise-compatible constructor value.
  Type 'TPromise<T | PromiseLike<T>>' is not assignable to type 'PromiseLike<T>'.
    Types of property 'then' are incompatible.
      Type '{ <U>(success?: (value: T | PromiseLike<T>) => TPromise<U>, error?: (err: any) => TPromise<U>, pr...' is not assignable to type '<TResult1 = T, TResult2 = never>(onfulfilled?: (value: T) => TResult1 | PromiseLike<TResult1>, on...'.
        Types of parameters 'success' and 'onfulfilled' are incompatible.
          Types of parameters 'value' and 'value' are incompatible.
            Type 'T | PromiseLike<T>' is not assignable to type 'T'.
              Type 'PromiseLike<T>' is not assignable to type 'T'.
[11:15:39] Error: /Users/jrieken/Code/vscode/src/vs/workbench/parts/extensions/electron-browser/extensionsViews.ts(119,29): Type 'typeof TPromise' is not a valid async function return type in ES5/ES3 because it does not refer to a Promise-compatible constructor value.
[11:15:39] Error: /Users/jrieken/Code/vscode/src/vs/workbench/parts/extensions/electron-browser/extensionsViews.ts(153,38): Type 'typeof TPromise' is not a valid async function return type in ES5/ES3 because it does not refer to a Promise-compatible constructor value.
[11:15:39] Error: /Users/jrieken/Code/vscode/src/vs/workbench/parts/extensions/electron-browser/extensionsViews.ts(429,29): Type 'typeof TPromise' is not a valid async function return type in ES5/ES3 because it does not refer to a Promise-compatible constructor value.
[11:15:39] Error: /Users/jrieken/Code/vscode/src/vs/workbench/parts/extensions/electron-browser/extensionsViews.ts(447,29): Type 'typeof TPromise' is not a valid async function return type in ES5/ES3 because it does not refer to a Promise-compatible constructor value.
[11:15:39] Error: /Users/jrieken/Code/vscode/src/vs/workbench/parts/extensions/electron-browser/extensionsViewlet.ts(303,28): Type 'typeof TPromise' is not a valid async function return type in ES5/ES3 because it does not refer to a Promise-compatible constructor value.
[11:15:39] Error: /Users/jrieken/Code/vscode/src/vs/platform/markers/common/problemMatcher.ts(1274,38): Argument of type 'ProblemPattern | ProblemPattern[]' is not assignable to parameter of type 'NamedMultiLineProblemPattern'.
  Type 'ProblemPattern' is not assignable to type 'NamedMultiLineProblemPattern'.
    Property 'name' is missing in type 'ProblemPattern'.
[11:15:39] Error: /Users/jrieken/Code/vscode/src/vs/platform/extensionManagement/node/extensionGalleryService.ts(358,44): Type 'typeof TPromise' is not a valid async function return type in ES5/ES3 because it does not refer to a Promise-compatible constructor value.
[11:15:39] Error: /Users/jrieken/Code/vscode/src/vs/editor/standalone/browser/standaloneLanguages.ts(703,2): Type '{ register: (language: ILanguageExtensionPoint) => void; getLanguages: () => ILanguageExtensionPo...' is not assignable to type 'typeof languages'.
  Types of property 'registerCodeActionProvider' are incompatible.
    Type '(languageId: string, provider: CodeActionProvider) => IDisposable' is not assignable to type '(languageId: string, provider: monaco.languages.CodeActionProvider) => monaco.IDisposable'.
      Types of parameters 'provider' and 'provider' are incompatible.
        Type 'monaco.languages.CodeActionProvider' is not assignable to type 'CodeActionProvider'.
          Types of property 'provideCodeActions' are incompatible.
            Type '(model: monaco.editor.IReadOnlyModel, range: monaco.Range, context: monaco.languages.CodeActionCo...' is not assignable to type '(model: IReadOnlyModel, range: Range, context: CodeActionContext, token: CancellationToken) => Co...'.
              Type 'Command[] | Thenable<Command[]>' is not assignable to type 'CodeAction[] | Thenable<CodeAction[]>'.
                Type 'Command[]' is not assignable to type 'CodeAction[] | Thenable<CodeAction[]>'.
                  Type 'Command[]' is not assignable to type 'Thenable<CodeAction[]>'.
                    Property 'then' is missing in type 'Command[]'.
[11:15:39] Error: /Users/jrieken/Code/vscode/src/vs/editor/common/standalone/standaloneBase.ts(229,2): Type '{ editor: undefined; languages: undefined; CancellationTokenSource: typeof CancellationTokenSourc...' is not assignable to type 'typeof monaco'.
  Types of property 'Promise' are incompatible.
    Type 'typeof TPromise' is not assignable to type 'typeof Promise'.
      Type 'TPromise<V | Thenable<V>>' is not assignable to type 'Promise<V>'.
        Types of property 'then' are incompatible.
          Type '{ <U>(success?: (value: V | Thenable<V>) => TPromise<U>, error?: (err: any) => TPromise<U>, progr...' is not assignable to type '{ <U>(success?: (value: V) => Promise<U>, error?: (err: any) => Promise<U>, progress?: ProgressCa...'.
            Types of parameters 'success' and 'success' are incompatible.
              Types of parameters 'value' and 'value' are incompatible.
                Type 'V | Thenable<V>' is not assignable to type 'V'.
                  Type 'Thenable<V>' is not assignable to type 'V'.
[11:15:39] Error: /Users/jrieken/Code/vscode/src/vs/workbench/electron-browser/extensionHost.ts(355,118): Type 'typeof TPromise' is not a valid async function return type in ES5/ES3 because it does not refer to a Promise-compatible constructor value.
[11:15:39] Error: /Users/jrieken/Code/vscode/extensions/git/src/askpass.ts(45,3): Type 'Promise<string | void>' is not assignable to type 'Promise<string>'.
  Type 'string | void' is not assignable to type 'string'.
    Type 'void' is not assignable to type 'string'.
[11:15:39] Finished compilation with 14 errors after 80248 ms
[11:15:39] Starting compilation...
[11:16:00] Error: /Users/jrieken/Code/vscode/src/vs/workbench/services/textfile/common/textFileService.ts(44,23): Class 'TextFileService' incorrectly implements interface 'ITextFileService'.
  Types of property 'revert' are incompatible.
    Type '(resource: URI, options?: IRevertOptions) => TPromise<boolean>' is not assignable to type '(resource: URI, force?: boolean) => TPromise<boolean>'.
      Types of parameters 'options' and 'force' are incompatible.
        Type 'boolean' is not assignable to type 'IRevertOptions'.
[11:16:00] Error: /Users/jrieken/Code/vscode/src/vs/workbench/services/textmodelResolver/test/textModelResolverService.test.ts(136,47): Type 'typeof TPromise' is not a valid async function return type in ES5/ES3 because it does not refer to a Promise-compatible constructor value.
[11:16:00] Error: /Users/jrieken/Code/vscode/src/vs/workbench/services/textmodelResolver/test/textModelResolverService.test.ts(136,47): Type 'typeof TPromise' is not a valid async function return type in ES5/ES3 because it does not refer to a Promise-compatible constructor value.
  Type 'TPromise<T | PromiseLike<T>>' is not assignable to type 'PromiseLike<T>'.
    Types of property 'then' are incompatible.
      Type '{ <U>(success?: (value: T | PromiseLike<T>) => TPromise<U>, error?: (err: any) => TPromise<U>, pr...' is not assignable to type '<TResult1 = T, TResult2 = never>(onfulfilled?: (value: T) => TResult1 | PromiseLike<TResult1>, on...'.
        Types of parameters 'success' and 'onfulfilled' are incompatible.
          Types of parameters 'value' and 'value' are incompatible.
            Type 'T | PromiseLike<T>' is not assignable to type 'T'.
              Type 'PromiseLike<T>' is not assignable to type 'T'.
[11:16:00] Error: /Users/jrieken/Code/vscode/src/vs/workbench/parts/extensions/electron-browser/extensionsViews.ts(119,29): Type 'typeof TPromise' is not a valid async function return type in ES5/ES3 because it does not refer to a Promise-compatible constructor value.
[11:16:00] Error: /Users/jrieken/Code/vscode/src/vs/workbench/parts/extensions/electron-browser/extensionsViews.ts(153,38): Type 'typeof TPromise' is not a valid async function return type in ES5/ES3 because it does not refer to a Promise-compatible constructor value.
[11:16:00] Error: /Users/jrieken/Code/vscode/src/vs/workbench/parts/extensions/electron-browser/extensionsViews.ts(429,29): Type 'typeof TPromise' is not a valid async function return type in ES5/ES3 because it does not refer to a Promise-compatible constructor value.
[11:16:00] Error: /Users/jrieken/Code/vscode/src/vs/workbench/parts/extensions/electron-browser/extensionsViews.ts(447,29): Type 'typeof TPromise' is not a valid async function return type in ES5/ES3 because it does not refer to a Promise-compatible constructor value.
[11:16:00] Error: /Users/jrieken/Code/vscode/src/vs/workbench/parts/extensions/electron-browser/extensionsViewlet.ts(303,28): Type 'typeof TPromise' is not a valid async function return type in ES5/ES3 because it does not refer to a Promise-compatible constructor value.
[11:16:00] Error: /Users/jrieken/Code/vscode/src/vs/platform/markers/common/problemMatcher.ts(1274,38): Argument of type 'ProblemPattern | ProblemPattern[]' is not assignable to parameter of type 'NamedMultiLineProblemPattern'.
  Type 'ProblemPattern' is not assignable to type 'NamedMultiLineProblemPattern'.
    Property 'name' is missing in type 'ProblemPattern'.
[11:16:00] Error: /Users/jrieken/Code/vscode/src/vs/platform/extensionManagement/node/extensionGalleryService.ts(358,44): Type 'typeof TPromise' is not a valid async function return type in ES5/ES3 because it does not refer to a Promise-compatible constructor value.
[11:16:00] Error: /Users/jrieken/Code/vscode/src/vs/workbench/electron-browser/extensionHost.ts(355,118): Type 'typeof TPromise' is not a valid async function return type in ES5/ES3 because it does not refer to a Promise-compatible constructor value.
[11:16:00] Error: /Users/jrieken/Code/vscode/extensions/git/src/askpass.ts(45,3): Type 'Promise<string | void>' is not assignable to type 'Promise<string>'.
  Type 'string | void' is not assignable to type 'string'.
    Type 'void' is not assignable to type 'string'.
[11:16:00] Finished compilation with 12 errors after 20429 ms
```

**Review by jrieken — COMMENTED** [review]
There still some formatting problem, compiler is happy tho. This the guide to run the formatter on all sources: https://github.com/Microsoft/vscode/pull/13373#issuecomment-252410642

**Pull request merged**